### PR TITLE
fix: upgrade cloud-init to v23.4.1 and add patch to retain exit code for recoverable errors

### PR DESCRIPTION
**Pull request by christopherco**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

Reverts the revert of cloud-init v23.4.1 and adds a patch backport to gracefully handle the new status exit code behavior which caused a regression in our extended testing apparatus, prompting the initial revert.

From https://bugs.launchpad.net/ubuntu/+source/cloud-init/+bug/2048522

`cloud-init status` introduced a new exit code of 2 with the meaning of recoverable errors / warnings.
Without this fix, status exited with 2 in cases where previously did it with 0, potentially breaking consumers (scripts or other programs) of `cloud-init status`.

The fix is to include a quilt patch retaining the exit code of 0 for recoverable errors / warnings for stable releases.

Fixes: 1a57d91875d208808582f483361b866fe0310453 ("Revert "fix: upgrade cloud-init to v23.4.1"")

Signed-off-by: Chris Co <chrco@microsoft.com>

###### Does this affect the toolchain?  

**NO**

###### Associated issues  

-  https://microsoft.visualstudio.com/OS/_workitems/edit/48381658

###### Test Methodology

Pipeline: https://dev.azure.com/mariner-org/mariner/_build/results?buildId=487581&view=results

Before backport (notice `echo $?` returns 2)
```
azureuser@chrco-marketplace-mariner2 [ ~ ]$ cloud-init --version
/usr/bin/cloud-init 23.4.1-1.cm2
azureuser@chrco-marketplace-mariner2 [ ~ ]$ cloud-init status --wait --format json
{
  "boot_status_code": "enabled-by-generator",
  "datasource": "azure",
  "detail": "DataSourceAzure [seed=/var/lib/waagent]",
  "errors": [],
  "extended_status": "degraded done",
  "init": {
    "errors": [],
    "finished": 1704960535.376083,
    "recoverable_errors": {},
    "start": 1704960535.169961
  },
  "init-local": {
    "errors": [],
    "finished": 1704960533.3723414,
    "recoverable_errors": {
      "ERROR": [
        "Azure datasource failure occurred: result=error|reason=failure to find primary DHCP interface|agent=Cloud-Init/23.4.1-1.cm2|interface=eth0|driver=hv_netvsc|router=10.0.0.1|stat
ic_routes=None|'lease={''interface'': ''eth0'', ''fixed-address'': ''10.0.0.4'', ''server-name'': ''IAD201121419001SOC'', ''subnet-mask'': ''255.255.255.0'', ''dhcp-lease-time'': ''4294
967295'', ''routers'': ''10.0.0.1'', ''dhcp-message-type'': ''5'', ''domain-name-servers'': ''168.63.129.16'', ''dhcp-server-identifier'': ''168.63.129.16'', ''dhcp-renewal-time'': ''42
94967295'', ''unknown-121'': ''0:a:0:0:1:20:a8:3f:81:10:a:0:0:1:20:a9:fe:a9:fe:a:0:0:1'', ''dhcp-rebinding-time'': ''4294967295'', ''unknown-245'': ''a8:3f:81:10'', ''domain-name'': ''s
immwbv3tw3ubfe412zalgj1wd.bx.internal.cloudapp.net'', ''renew'': ''0 2160/02/17 14:37:08'', ''rebind'': ''0 2160/02/17 14:37:08'', ''expire'': ''0 2160/02/17 14:37:08''}'|vm_id=9ed7f284
-ae3d-430a-8b22-722c8ba878ab|timestamp=2024-01-11T08:08:53.261840|documentation_url=https://aka.ms/linuxprovisioningerror"
      ]
    },
    "start": 1704960533.1129274
  },
  "last_update": "Thu, 11 Jan 2024 08:08:55 +0000",
  "modules-config": {
    "errors": [],
    "finished": 1704960535.6374042,
    "recoverable_errors": {},
    "start": 1704960535.5896227
  },
  "modules-final": {
    "errors": [],
    "finished": 1704960535.8863187,
    "recoverable_errors": {},
    "start": 1704960535.8338778
  },
  "recoverable_errors": {
    "ERROR": [
      "Azure datasource failure occurred: result=error|reason=failure to find primary DHCP interface|agent=Cloud-Init/23.4.1-1.cm2|interface=eth0|driver=hv_netvsc|router=10.0.0.1|static
_routes=None|'lease={''interface'': ''eth0'', ''fixed-address'': ''10.0.0.4'', ''server-name'': ''IAD201121419001SOC'', ''subnet-mask'': ''255.255.255.0'', ''dhcp-lease-time'': ''429496
7295'', ''routers'': ''10.0.0.1'', ''dhcp-message-type'': ''5'', ''domain-name-servers'': ''168.63.129.16'', ''dhcp-server-identifier'': ''168.63.129.16'', ''dhcp-renewal-time'': ''4294
967295'', ''unknown-121'': ''0:a:0:0:1:20:a8:3f:81:10:a:0:0:1:20:a9:fe:a9:fe:a:0:0:1'', ''dhcp-rebinding-time'': ''4294967295'', ''unknown-245'': ''a8:3f:81:10'', ''domain-name'': ''sim
mwbv3tw3ubfe412zalgj1wd.bx.internal.cloudapp.net'', ''renew'': ''0 2160/02/17 14:37:08'', ''rebind'': ''0 2160/02/17 14:37:08'', ''expire'': ''0 2160/02/17 14:37:08''}'|vm_id=9ed7f284-a
e3d-430a-8b22-722c8ba878ab|timestamp=2024-01-11T08:08:53.261840|documentation_url=https://aka.ms/linuxprovisioningerror"
    ]
  },
  "stage": null,
  "status": "done"
}
azureuser@chrco-marketplace-mariner2 [ ~ ]$ echo $?
2
```
After backport (notice `echo $?` now returns 0)
```
azureuser@chrco-marketplace-mariner2 [ ~ ]$ cloud-init --version
/usr/bin/cloud-init 23.4.1-2.cm2
azureuser@chrco-marketplace-mariner2 [ ~ ]$ cloud-init status --wait --format json
{
  "boot_status_code": "enabled-by-generator",
  "datasource": "azure",
  "detail": "DataSourceAzure [seed=/var/lib/waagent]",
  "errors": [],
  "extended_status": "degraded done",
  "init": {
    "errors": [],
    "finished": 1704961425.7507293,
    "recoverable_errors": {},
    "start": 1704961425.406041
  },
  "init-local": {
    "errors": [],
    "finished": 1704961423.353731,
    "recoverable_errors": {
      "ERROR": [
        "Azure datasource failure occurred: result=error|reason=failure to find primary DHCP interface|agent=Cloud-Init/23.4.1-2.cm2|interface=eth0|driver=hv_netvsc|router=10.0.0.1|static_routes=None|'lease={''interface'': ''eth0'', ''fixed-address'': ''10.0.0.4'', ''server-name'': ''IAD201121419001SOC'', ''subnet-mask'': ''255.255.255.0'', ''dhcp-lease-time'': ''4294967295'', ''routers'': ''10.0.0.1'', ''dhcp-message-type'': ''5'', ''domain-name-servers'': ''168.63.129.16'', ''dhcp-server-identifier'': ''168.63.129.16'', ''dhcp-renewal-time'': ''4294967295'', ''unknown-121'': ''0:a:0:0:1:20:a8:3f:81:10:a:0:0:1:20:a9:fe:a9:fe:a:0:0:1'', ''dhcp-rebinding-time'': ''4294967295'', ''unknown-245'': ''a8:3f:81:10'', ''domain-name'': ''simmwbv3tw3ubfe412zalgj1wd.bx.internal.cloudapp.net'', ''renew'': ''0 2160/02/17 14:51:57'', ''rebind'': ''0 2160/02/17 14:51:57'', ''expire'': ''0 2160/02/17 14:51:57''}'|vm_id=9ed7f284-ae3d-430a-8b22-722c8ba878ab|timestamp=2024-01-11T08:23:43.006497|documentation_url=https://aka.ms/linuxprovisioningerror"
      ]
    },
    "start": 1704961422.5247135
  },
  "last_update": "Thu, 11 Jan 2024 08:23:46 +0000",
  "modules-config": {
    "errors": [],
    "finished": 1704961426.0113945,
    "recoverable_errors": {},
    "start": 1704961425.969099
  },
  "modules-final": {
    "errors": [],
    "finished": 1704961426.2613358,
    "recoverable_errors": {},
    "start": 1704961426.207721
  },
  "recoverable_errors": {
    "ERROR": [
      "Azure datasource failure occurred: result=error|reason=failure to find primary DHCP interface|agent=Cloud-Init/23.4.1-2.cm2|interface=eth0|driver=hv_netvsc|router=10.0.0.1|static_routes=None|'lease={''interface'': ''eth0'', ''fixed-address'': ''10.0.0.4'', ''server-name'': ''IAD201121419001SOC'', ''subnet-mask'': ''255.255.255.0'', ''dhcp-lease-time'': ''4294967295'', ''routers'': ''10.0.0.1'', ''dhcp-message-type'': ''5'', ''domain-name-servers'': ''168.63.129.16'', ''dhcp-server-identifier'': ''168.63.129.16'', ''dhcp-renewal-time'': ''4294967295'', ''unknown-121'': ''0:a:0:0:1:20:a8:3f:81:10:a:0:0:1:20:a9:fe:a9:fe:a:0:0:1'', ''dhcp-rebinding-time'': ''4294967295'', ''unknown-245'': ''a8:3f:81:10'', ''domain-name'': ''simmwbv3tw3ubfe412zalgj1wd.bx.internal.cloudapp.net'', ''renew'': ''0 2160/02/17 14:51:57'', ''rebind'': ''0 2160/02/17 14:51:57'', ''expire'': ''0 2160/02/17 14:51:57''}'|vm_id=9ed7f284-ae3d-430a-8b22-722c8ba878ab|timestamp=2024-01-11T08:23:43.006497|documentation_url=https://aka.ms/linuxprovisioningerror"
    ]
  },
  "stage": null,
  "status": "done"
}
azureuser@chrco-marketplace-mariner2 [ ~ ]$ echo $?
0
```

### CODE REVIEW

File: SPECS/cloud-init/cloud-init.spec [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
+Release:        2%{?dist}
 License:        GPLv3
 Vendor:         Microsoft Corporation
 Distribution:   Mariner
 Group:          System Environment/Base
 URL:            https://launchpad.net/cloud-init
 Source0:        https://launchpad.net/cloud-init/trunk/%{version}/+download/%{name}-%{version}.tar.gz
 Source1:        10-azure-kvp.cfg
-Patch0:         overrideDatasourceDetection.patch
+Patch0:         Retain-exit-code-in-cloud-init-status-for-recoverabl.patch
```
Comment (rmhsawyer) [inline SPECS/cloud-init/cloud-init.spec:12]: nit in patch's name: recoverabl -> recoverable

**Review by rmhsawyer — APPROVED** [review]
Can you please add provisioning tests if not done yet:
1. Azure VM/AKS provisioning with new cloud-init package
2. local hyper-v provisioning

**Comment by christopherco** [comment]
AKS tests pass - https://dev.azure.com/mariner-org/mariner/_build/results?buildId=487399&view=results

File: SPECS/cloud-init/cloud-init.spec (hunk 1 above)
Comment (christopherco) [reply SPECS/cloud-init/cloud-init.spec:12]: This file name comes from the output of `git format-patch` (minus the initial numbering). Will keep this naming generation consistent

**Comment by christopherco** [comment]
AKS tests pass after update:
- https://dev.azure.com/mariner-org/mariner/_build/results?buildId=492179&view=results
- https://dev.azure.com/mariner-org/mariner/_build/results?buildId=492180&view=results

**Pull request merged**